syntax = "proto3";

import "proto/blackbox_variable.proto";

enum Algorithm {
    RANDOM_SEARCH = 0;
    SIMULATED_ANNEALING = 1;
    NELDER_MEAD = 2;
    OPENTUNER = 3;
}

message EvaluationJob {
  string evaluation_job_id = 1;
  repeated BlackboxVariable variables = 2;
  repeated BlackboxVariableMetadata variables_metadata = 3;
}

message OptimizationJob {
  message JobParameters {
    float time_limit = 1;
    int64 max_evaluations = 2;
  }
  message JobConstraint {
    message VariableConstraint {
      string name = 1;
      float coefficient = 2;
    }
    repeated VariableConstraint variable_constraints = 1;
    float constant_term = 2;
  }
  repeated BlackboxVariableMetadata task_variables = 1;
  Algorithm algorithm = 2;
  JobParameters job_parameters = 3;
  repeated JobConstraint job_constraints = 4;
}

message Job {
  string job_id = 1;
  oneof job_type {
    EvaluationJob evaluation_job = 3;
    OptimizationJob optimization_job = 4;
  }
}

message CompletedEvaluationJob {
  double result = 1;
}

message CompletedOptimizationJob {
  double result = 1;
  repeated BlackboxVariable variables = 2;
}

message CompletedJob {
  string job_id = 1;
  string task_id = 2;
  oneof job_type {
    CompletedEvaluationJob evaluation_job = 3;
    CompletedOptimizationJob optimization_job = 4;
  }
}
